@bors r-
 
 2024-08-12T08:09:07.8523294Z [1m[32m   Compiling[0m clap_derive v4.5.5
2024-08-12T08:09:08.7043836Z [RUSTC-TIMING] tracing_attributes test:false 1.637
2024-08-12T08:09:08.7050616Z [1m[32m   Compiling[0m thiserror-impl v1.0.61
2024-08-12T08:09:10.1303074Z [RUSTC-TIMING] thiserror_impl test:false 1.419
2024-08-12T08:09:10.1792105Z [RUSTC-TIMING] thiserror test:false 0.043
2024-08-12T08:09:10.1798434Z [1m[32m   Compiling[0m tracing v0.1.37
2024-08-12T08:09:10.2561947Z [RUSTC-TIMING] clap_derive test:false 2.397
2024-08-12T08:09:10.2568760Z [1m[32m   Compiling[0m clap v4.5.7
2024-08-12T08:09:10.2983590Z [RUSTC-TIMING] clap test:false 0.034
2024-08-12T08:09:10.5155182Z [RUSTC-TIMING] tracing test:false 0.330
2024-08-12T08:09:10.5173792Z [1m[32m   Compiling[0m llvm-bitcode-linker v0.0.1 (/checkout/src/tools/llvm-bitcode-linker)
2024-08-12T08:09:11.0202785Z [RUSTC-TIMING] llvm_bitcode_linker test:false 0.498
2024-08-12T08:09:11.1082040Z [RUSTC-TIMING] clap_builder test:false 8.264
2024-08-12T08:09:12.3069458Z [RUSTC-TIMING] llvm_bitcode_linker test:false 1.194
2024-08-12T08:09:12.3083923Z [1m[32m    Finished[0m `release` profile [optimized] target(s) in 11.09s
2024-08-12T08:09:12.3229000Z ##[endgroup]
2024-08-12T08:09:12.3230751Z [TIMING] core::build_steps::tool::LlvmBitcodeLinker { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu }, target: x86_64-unknown-linux-gnu, extra_features: [] } -- 11.115
2024-08-12T08:09:12.3233162Z [TIMING] core::build_steps::compile::Assemble { target_compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu } } -- 0.003
2024-08-12T08:09:12.3234606Z [TIMING] core::build_steps::compile::StartupObjects { compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu }, target: loongarch64-unknown-linux-musl } -- 0.000
2024-08-12T08:09:12.3236187Z [TIMING] core::builder::Builder::sysroot_libdir::Libdir { compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu }, target: loongarch64-unknown-linux-musl } -- 0.000
2024-08-12T08:09:12.3278104Z thread 'main' panicked at src/lib.rs:1794:13:
2024-08-12T08:09:12.3279076Z set_file_times(dst, file_times) failed with Permission denied (os error 13)
2024-08-12T08:09:12.3279792Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-08-12T08:09:12.3301323Z Build completed unsuccessfully in 0:16:13
2024-08-12T08:09:12.3381725Z   local time: Mon Aug 12 08:09:12 UTC 2024
2024-08-12T08:09:12.4135264Z   network time: Mon, 12 Aug 2024 08:09:12 GMT
2024-08-12T08:09:12.6295762Z ##[error]Process completed with exit code 1.
2024-08-12T08:09:12.6402805Z Post job cleanup.

--- a/src/bootstrap/src/core/download.rs
+++ b/src/bootstrap/src/core/download.rs
@@ -703,9 +703,7 @@ download-rustc = false
             let file_times = fs::FileTimes::new().set_accessed(now).set_modified(now);
 
             let llvm_config = llvm_root.join("bin").join(exe("llvm-config", self.build));
-            let llvm_config_file = t!(File::options().write(true).open(llvm_config));
-
-            t!(llvm_config_file.set_times(file_times));
+            t!(crate::utils::helpers::set_file_times(llvm_config, file_times));
 
             if self.should_fix_bins_and_dylibs() {
                 let llvm_lib = llvm_root.join("lib");

--- a/src/bootstrap/src/lib.rs
+++ b/src/bootstrap/src/lib.rs
@@ -37,7 +37,9 @@ use crate::core::builder;
 use crate::core::builder::{Builder, Kind};
 use crate::core::config::{flags, DryRun, LldMode, LlvmLibunwind, Target, TargetSelection};
 use crate::utils::exec::{command, BehaviorOnFailure, BootstrapCommand, CommandOutput, OutputMode};
-use crate::utils::helpers::{self, dir_is_empty, exe, libdir, mtime, output, symlink_dir};
+use crate::utils::helpers::{
+    self, dir_is_empty, exe, libdir, mtime, output, set_file_times, symlink_dir,
+};
 
 mod core;
 mod utils;
@@ -1792,21 +1794,20 @@ Executed at: {executed_at}"#,
             }
         }
         if let Ok(()) = fs::hard_link(&src, dst) {
-            // Attempt to "easy copy" by creating a hard link
-            // (symlinks don't work on windows), but if that fails
-            // just fall back to a slow `copy` operation.
+            // Attempt to "easy copy" by creating a hard link (symlinks are priviledged on windows),
+            // but if that fails just fall back to a slow `copy` operation.
         } else {
             if let Err(e) = fs::copy(&src, dst) {
                 panic!("failed to copy `{}` to `{}`: {}", src.display(), dst.display(), e)
             }
             t!(fs::set_permissions(dst, metadata.permissions()));
 
+            // Restore file times because changing permissions on e.g. Linux using `chmod` can cause
+            // file access time to change.
             let file_times = fs::FileTimes::new()
                 .set_accessed(t!(metadata.accessed()))
                 .set_modified(t!(metadata.modified()));
-
-            let dst_file = t!(fs::File::open(dst));
-            t!(dst_file.set_times(file_times));
+            t!(set_file_times(dst, file_times));
         }
     }
 

--- a/src/bootstrap/src/utils/helpers.rs
+++ b/src/bootstrap/src/utils/helpers.rs
@@ -544,3 +544,15 @@ pub fn get_closest_merge_base_commit(
 
     Ok(output_result(git.as_command_mut())?.trim().to_owned())
 }
+
+/// Sets the file times for a given file at `path`.
+pub fn set_file_times<P: AsRef<Path>>(path: P, times: fs::FileTimes) -> io::Result<()> {
+    // Windows requires file to be writable to modify file times. But on Linux CI the file does not
+    // need to be writable to modify file times and might be read-only.
+    let f = if cfg!(windows) {
+        fs::File::options().write(true).open(path)?
+    } else {
+        fs::File::open(path)?
+    };
+    f.set_times(times)
+}

--- a/src/bootstrap/src/utils/helpers/tests.rs
+++ b/src/bootstrap/src/utils/helpers/tests.rs
@@ -3,7 +3,8 @@ use std::io::Write;
 use std::path::PathBuf;
 
 use crate::utils::helpers::{
-    check_cfg_arg, extract_beta_rev, hex_encode, make, program_out_of_date, symlink_dir,
+    check_cfg_arg, extract_beta_rev, hex_encode, make, program_out_of_date, set_file_times,
+    symlink_dir,
 };
 use crate::{Config, Flags};
 
@@ -91,4 +92,26 @@ fn test_symlink_dir() {
     fs::remove_dir(link_path).unwrap();
     #[cfg(not(windows))]
     fs::remove_file(link_path).unwrap();
+}
+
+#[test]
+fn test_set_file_times_sanity_check() {
+    let config =
+        Config::parse(Flags::parse(&["check".to_owned(), "--config=/does/not/exist".to_owned()]));
+    let tempfile = config.tempdir().join(".tmp-file");
+
+    {
+        File::create(&tempfile).unwrap().write_all(b"dummy value").unwrap();
+        assert!(tempfile.exists());
+    }
+
+    // This might only fail on Windows (if file is default read-only then we try to modify file
+    // times).
+    let unix_epoch = std::time::SystemTime::UNIX_EPOCH;
+    let target_time = fs::FileTimes::new().set_accessed(unix_epoch).set_modified(unix_epoch);
+    set_file_times(&tempfile, target_time).unwrap();
+
+    let found_metadata = fs::metadata(tempfile).unwrap();
+    assert_eq!(found_metadata.accessed().unwrap(), unix_epoch);
+    assert_eq!(found_metadata.modified().unwrap(), unix_epoch)
 }

